devdoc/thandle_tuple_array_requirements.md

```diff
@@ -0,0 +1,84 @@
+`thandle_tuple_array` requirements
+================
+
+## Overview
+
+`thandle_tuple_array` holds an array of structs with `THANDLE(<TYPE>)` members along with the count of array elements. That is, an array of tuples where each tuple field is a `THANDLE`.
+
+## Exposed API
+
+```c
+#define DECLARE_THANDLE_TUPLE_TYPE(name, ...) \
+    ...
+
+#define DECLARE_THANDLE_TUPLE_ARRAY_TYPE(name) \
+    ...
+
+#define THANDLE_TUPLE_ARRAY_CREATE(name) ...
+#define THANDLE_TUPLE_ARRAY_DESTROY(name) ...
+
+#define DECLARE_THANDLE_TUPLE_ARRAY(name, ...) \
+    DECLARE_THANDLE_TUPLE_TYPE(name, __VAR_ARGS__) \
+    DECLARE_THANDLE_TUPLE_ARRAY_TYPE(name) \
+    MOCKABLE_FUNCTION(, THANDLE_TUPLE_ARRAY_TYPE(name)*, THANDLE_TUPLE_ARRAY_CREATE(name), uint32_t, count); \
+    MOCKABLE_FUNCTION(, void, THANDLE_TUPLE_ARRAY_DESTROY(name), THANDLE_TUPLE_ARRAY_TYPE(name)*, tuple_array);
+
+#define DEFINE_THANDLE_TUPLE_ARRAY(name, ...) \
+    ...
+```
+
+### Example Usage
+
+```c
+// In the header
+#define MY_TUPLE_FIELDS \
+    RC_STRING, foo, \
+    RC_STRING, bar, \
+    RC_STRING, baz
+
+DECLARE_THANDLE_TUPLE_ARRAY(MY_TUPLE, MY_TUPLE_FIELDS);
+
+// In .c file:
+DEFINE_THANDLE_TUPLE_ARRAY(MY_TUPLE, MY_TUPLE_FIELDS)
+
+// Callers
+THANDLE_TUPLE_ARRAY_TYPE(MY_TUPLE) tuple = THANDLE_TUPLE_ARRAY_CREATE(MY_TUPLE)(4);

```

Comment:
>tuple [](http://example.com/codeflow?start=35&length=5)

is THANDLE_INC_REF/DEC_REF working for tuple here? Would be nice with an example.

Reply:
Truth being said, documentation doesn't say anything about that.

So these arrays are not refcounted, why are they called with THANDLE in their name? I assumed they'd be refcounted (still unsure) by the name only.

Reply:
Hmm yeah maybe just `TUPLE_ARRAY` is the analogous name to `rc_string_array`, since the THANDLE's are just used for the fields...
Though it does hold THANDLE's (and maybe it's good to have that called out in the name), it's just not a THANDLE itself...

Reply:
Added the note that this is not ref counted. I'm leaving the name because I want it clear to callers that tuple member access is with THANDLE's.